lib/puppet/provider/cisco_itd_device_group/cisco.rb

```diff
@@ -0,0 +1,200 @@
+rescue LoadError # seen on master, not on agent
+  # See longstanding Puppet issues #4248, #7316, #14073, #14149, etc. Ugh.
+  require File.expand_path(File.join(File.dirname(__FILE__), '..', '..', '..',
+                                     'puppet_x', 'cisco', 'cmnutils.rb'))
+end
+Puppet::Type.type(:cisco_itd_device_group).provide(:cisco) do
+  desc 'The Cisco provider for cisco_itd_device_group.'
+
+  confine feature: :cisco_node_utils
+  defaultfor operatingsystem: :nexus
+
+  mk_resource_methods
+
+  ITDDG_NON_BOOL_PROPS = [
+    :probe_type,
+    :probe_dns_host,
+    :probe_frequency,
+    :probe_port,
+    :probe_retry_down,
+    :probe_retry_up,
+    :probe_timeout,
+  ]
+  ITDDG_BOOL_PROPS = [
+    :probe_control
+  ]
+  ITDDG_ALL_PROPS = ITDDG_NON_BOOL_PROPS + ITDDG_BOOL_PROPS
+
+  PuppetX::Cisco::AutoGen.mk_puppet_methods(:non_bool, self, '@nu',
+                                            ITDDG_NON_BOOL_PROPS)
+  PuppetX::Cisco::AutoGen.mk_puppet_methods(:bool, self, '@nu',
+                                            ITDDG_BOOL_PROPS)
+
+  def initialize(value={})
+    super(value)
+    @nu = Cisco::ItdDeviceGroup.itds[@property_hash[:name]]
+    @property_flush = {}
+  end
+
+  def self.properties_get(itd_device_group_name, nu_obj)
+    debug "Checking instance, #{itd_device_group_name}."
+    current_state = {
+      itddg:  itd_device_group_name,
+      name:   itd_device_group_name,
+      ensure: :present,
+    }
+    # Call node_utils getter for each property
+    ITDDG_NON_BOOL_PROPS.each do |prop|
+      current_state[prop] = nu_obj.send(prop)
+    end
+
+    ITDDG_BOOL_PROPS.each do |prop|
+      val = nu_obj.send(prop)
+      if val.nil?
+        current_state[prop] = nil
+      else
+        current_state[prop] = val ? :true : :false
+      end
+    end
+    new(current_state)
+  end # self.properties_get
+
+  def self.instances
+    itds = []
+    Cisco::ItdDeviceGroup.itds.each do |itd_device_group_name, nu_obj|
+      itds << properties_get(itd_device_group_name, nu_obj)
+    end
+    itds
+  end
+
+  def self.prefetch(resources)
+    itds = instances
+    resources.keys.each do |name|
+      provider = itds.find { |nu_obj| nu_obj.instance_name == name }
+      resources[name].provider = provider unless provider.nil?
+    end
+  end # self.prefetch
+
+  def exists?
+    (@property_hash[:ensure] == :present)
+  end
+
+  def create
+    @property_flush[:ensure] = :present
+  end
+
+  def destroy
+    @property_flush[:ensure] = :absent
+  end
+
+  def instance_name
+    itddg
+  end
+
+  def properties_set(new_itd=false)
+    ITDDG_ALL_PROPS.each do |prop|
+      next unless @resource[prop]
+      send("#{prop}=", @resource[prop]) if new_itd
+      unless @property_flush[prop].nil?
+        @nu.send("#{prop}=", @property_flush[prop]) if
+          @nu.respond_to?("#{prop}=")
+      end
+    end
+    # custom setters which require one-shot multi-param setters
+    probe_set
+  end
+
+  def fail_attribute_check(type)
```

Comment:
If possible, it's better to move this kind of validation into the type.

Example: https://github.com/cisco/cisco-network-puppet-module/blob/develop/lib/puppet/type/cisco_vxlan_vtep.rb#L123